// nx-stylelint@18.0.2/src/generators/configuration/generator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.configurationGenerator = configurationGenerator;
const devkit_1 = require("@nx/devkit");
const config_file_1 = require("../../utils/config-file");
const formatter_1 = require("../../utils/formatter");
const generator_1 = require("../init/generator");
async function configurationGenerator(host, options) {
    const init = await (0, generator_1.initGenerator)(host, { skipFormat: options.skipFormat, scss: options.scss });
    const normalizedOptions = normalizeSchema(host, options);
    if (normalizedOptions.stylelintTargetExists) {
        devkit_1.logger.error(`Project '${options.project}' already has a stylelint target.`);
        return;
    }
    devkit_1.logger.info(`Adding Stylelint configuration and target to '${options.project}' ...\n`);
    (0, config_file_1.createProjectStylelintConfigFile)(host, normalizedOptions.projectRoot, normalizedOptions.scss);
    addTarget(host, normalizedOptions);
    if (options.skipFormat !== true)
        await (0, devkit_1.formatFiles)(host);
    return init;
}
exports.default = configurationGenerator;
function normalizeSchema(tree, options) {
    const projectConfig = (0, devkit_1.readProjectConfiguration)(tree, options.project);
    let formatter = formatter_1.defaultFormatter;
    if ((0, formatter_1.isCoreFormatter)(options.formatter)) {
        formatter = options.formatter;
    }
    else if (options.formatter) {
        devkit_1.logger.error(`Given formatter '${options.formatter}' is not a stylelint core formatter. Falling back to '${formatter_1.defaultFormatter}' formatter.`);
    }
    return {
        project: options.project,
        formatter,
        scss: !!options.scss,
        skipFormat: !!options.skipFormat,
        projectRoot: projectConfig.root,
        stylelintTargetExists: !!projectConfig.targets?.['stylelint'],
    };
}
function addTarget(tree, options) {
    const projectConfig = (0, devkit_1.readProjectConfiguration)(tree, options.project);
    const targetOptions = {
        lintFilePatterns: [(0, devkit_1.joinPathFragments)(options.projectRoot, '**', '*.css')],
        formatter: options.formatter === 'string' ? undefined : options.formatter,
    };
    if (options.scss)
        targetOptions.lintFilePatterns.push((0, devkit_1.joinPathFragments)(options.projectRoot, '**', '*.scss'));
    projectConfig.targets = {
        ...projectConfig.targets,
        stylelint: {
            executor: 'nx-stylelint:lint',
            outputs: ['{options.outputFile}'],
            options: targetOptions,
        },
    };
    (0, devkit_1.updateProjectConfiguration)(tree, options.project, projectConfig);
}
