// nx-stylelint@18.0.2/src/utils/config-file.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createRootStylelintConfigFile = exports.STYLELINT_CONFIG_FILE = exports.STYLELINT_CONFIG_FILE_PATTERN = void 0;
exports.createProjectStylelintConfigFile = createProjectStylelintConfigFile;
exports.isCompatibleRootConfig = isCompatibleRootConfig;
exports.getInputConfigFiles = getInputConfigFiles;
exports.readAffectingStylelintConfigFiles = readAffectingStylelintConfigFiles;
const devkit_1 = require("@nx/devkit");
const node_fs_1 = require("node:fs");
const node_path_1 = require("node:path");
const config_1 = require("./config");
const path_1 = require("./path");
const cosmiconfig_1 = require("./cosmiconfig");
exports.STYLELINT_CONFIG_FILE_PATTERN = '.stylelintrc(.(json|yml|yaml|js))?';
exports.STYLELINT_CONFIG_FILE = '.stylelintrc.json';
const createRootStylelintConfigFile = (tree, withScssSupport) => (0, devkit_1.writeJson)(tree, exports.STYLELINT_CONFIG_FILE, withScssSupport ? config_1.ROOT_STYLELINT_CONFIG_SCSS : config_1.ROOT_STYLELINT_CONFIG);
exports.createRootStylelintConfigFile = createRootStylelintConfigFile;
function createProjectStylelintConfigFile(tree, projectRoot, withScssSupport) {
    const config = {
        extends: [(0, devkit_1.joinPathFragments)((0, devkit_1.offsetFromRoot)(projectRoot), exports.STYLELINT_CONFIG_FILE)],
        ignoreFiles: ['!**/*'],
        overrides: [
            {
                files: ['**/*.css'],
                rules: {},
            },
        ],
    };
    if (withScssSupport) {
        config.overrides.push(config_1.PROJECT_STYLELINT_CONFIG_SCSS_OVERRIDE);
    }
    (0, devkit_1.writeJson)(tree, (0, devkit_1.joinPathFragments)(projectRoot, exports.STYLELINT_CONFIG_FILE), config);
}
function isCompatibleRootConfig(tree) {
    if (!tree.exists(exports.STYLELINT_CONFIG_FILE))
        return false;
    const config = (0, devkit_1.readJson)(tree, exports.STYLELINT_CONFIG_FILE);
    return config.ignoreFiles === '**/*' || (Array.isArray(config.ignoreFiles) && config.ignoreFiles.includes('**/*'));
}
async function getInputConfigFiles(configFilePath, projectRoot) {
    return [...(await readAffectingStylelintConfigFiles(configFilePath))]
        .map((configFilePath) => {
        if (configFilePath.startsWith(devkit_1.workspaceRoot)) {
            configFilePath = (0, node_path_1.relative)(devkit_1.workspaceRoot, configFilePath);
            if (configFilePath.startsWith(projectRoot)) {
                configFilePath = (0, devkit_1.joinPathFragments)('{projectRoot}', configFilePath.substring(projectRoot.length));
            }
            else {
                configFilePath = (0, devkit_1.joinPathFragments)('{workspaceRoot}', configFilePath);
            }
        }
        return configFilePath;
    })
        .sort();
}
async function readAffectingStylelintConfigFiles(configFilePath) {
    if (!(0, node_fs_1.existsSync)(configFilePath))
        return new Set();
    try {
        const result = await (0, cosmiconfig_1.loadStylelintConfig)(configFilePath);
        if (!result)
            return new Set();
        const stylelintConfigFiles = new Set([result.filepath]);
        const stylelintConfig = result.config;
        if (stylelintConfig && (stylelintConfig.extends || stylelintConfig.overrides)) {
            const extendsItems = new Set(stylelintConfig.extends ? [stylelintConfig.extends].flat().filter((v) => typeof v === 'string') : []);
            if (stylelintConfig.overrides) {
                for (const override of stylelintConfig.overrides) {
                    if (override.extends) {
                        for (const value of [override.extends].flat().filter((v) => typeof v === 'string')) {
                            extendsItems.add(value);
                        }
                    }
                }
            }
            if (extendsItems.size > 0) {
                const readAndAdd = async (path) => {
                    for (const value of await readAffectingStylelintConfigFiles(path)) {
                        stylelintConfigFiles.add(value);
                    }
                };
                for (const extendFilePath of extendsItems) {
                    if ((0, node_path_1.isAbsolute)(extendFilePath)) {
                        await readAndAdd(extendFilePath);
                    }
                    else if ((0, path_1.isRelativePath)(extendFilePath)) {
                        await readAndAdd((0, node_path_1.join)((0, node_path_1.dirname)(configFilePath), extendFilePath));
                    }
                }
            }
        }
        return stylelintConfigFiles;
    }
    catch (err) {
        console.error(err);
        throw err;
    }
}
