// nx-stylelint@18.0.2/src/plugins/plugin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createNodes = exports.createNodesV2 = void 0;
const devkit_1 = require("@nx/devkit");
const calculate_hash_for_create_nodes_1 = require("@nx/devkit/src/utils/calculate-hash-for-create-nodes");
const node_fs_1 = require("node:fs");
const nodePath = require("node:path");
const config_file_1 = require("../utils/config-file");
const workspace_data_directory_1 = require("../utils/workspace-data-directory");
const pmc = (0, devkit_1.getPackageManagerCommand)();
const STYLELINT_CONFIG_FILES = [
    '.stylelintrc',
    '.stylelintrc.json',
    '.stylelintrc.yml',
    '.stylelintrc.yaml',
    '.stylelintrc.js',
    '.stylelintrc.cjs',
    '.stylelintrc.mjs',
    '.stylelintrc.ts',
    '.stylelintrc.cts',
    '.stylelintrc.mts',
    'stylelint.config.js',
    'stylelint.config.cjs',
    'stylelint.config.mjs',
    'stylelint.config.ts',
    'stylelint.config.cts',
    'stylelint.config.mts',
];
const STYLELINT_CONFIG_FILES_GLOB = `{${STYLELINT_CONFIG_FILES.map((f) => `**/${f}`).join(',')}}`;
function readTargetsCache(cachePath) {
    return (0, node_fs_1.existsSync)(cachePath) ? (0, devkit_1.readJsonFile)(cachePath) : {};
}
function writeTargetsToCache(cachePath, results) {
    (0, devkit_1.writeJsonFile)(cachePath, results);
}
exports.createNodesV2 = [
    STYLELINT_CONFIG_FILES_GLOB,
    async (projectConfigurationFiles, options, context) => {
        const normalizedOptions = normalizeOptions(options);
        const optionsHash = (0, devkit_1.hashArray)(Object.entries(normalizedOptions).map(([key, value]) => key + JSON.stringify(value)));
        const cachePath = nodePath.join(workspace_data_directory_1.workspaceDataDirectory, `stylelint-${optionsHash}.hash`);
        const targetsCache = readTargetsCache(cachePath);
        try {
            return await (0, devkit_1.createNodesFromFiles)((configFile, _, context) => createNodesInternal(configFile, normalizedOptions, context, targetsCache), projectConfigurationFiles, normalizedOptions, context);
        }
        finally {
            writeTargetsToCache(cachePath, targetsCache);
        }
    },
];
exports.createNodes = [
    STYLELINT_CONFIG_FILES_GLOB,
    async (configFilePath, options, context) => {
        devkit_1.logger.warn('`createNodes` is deprecated. Update your plugin to utilize createNodesV2 instead. In Nx 20, this will change to the createNodesV2 API.');
        return createNodesInternal(configFilePath, normalizeOptions(options), context, {});
    },
];
async function createNodesInternal(configFilePath, options, context, targetsCache) {
    const projectRoot = nodePath.dirname(configFilePath);
    const siblingFiles = (0, node_fs_1.readdirSync)(nodePath.join(context.workspaceRoot, projectRoot));
    if (!siblingFiles.includes('package.json') && !siblingFiles.includes('project.json')) {
        return {};
    }
    const isStandaloneWorkspace = projectRoot === '.' &&
        (0, node_fs_1.existsSync)(nodePath.join(context.workspaceRoot, projectRoot, 'src')) &&
        (0, node_fs_1.existsSync)(nodePath.join(context.workspaceRoot, projectRoot, 'package.json'));
    if (projectRoot === '.' && !isStandaloneWorkspace)
        return {};
    const hash = await (0, calculate_hash_for_create_nodes_1.calculateHashForCreateNodes)(projectRoot, options, context);
    targetsCache[hash] ?? (targetsCache[hash] = await stylelintTarget(configFilePath, projectRoot, options));
    const target = targetsCache[hash];
    const project = {
        root: projectRoot,
        targets: {
            [options.targetName]: target,
        },
    };
    return {
        projects: {
            [projectRoot]: project,
        },
    };
}
async function stylelintTarget(configFilePath, projectRoot, options) {
    const inputConfigFiles = await (0, config_file_1.getInputConfigFiles)(configFilePath, projectRoot);
    const glob = options.extensions.length === 1 ? `**/*.${options.extensions[0]}` : `**/*.{${options.extensions.join(',')}}`;
    return {
        command: `stylelint "${glob}"`,
        cache: true,
        options: {
            cwd: projectRoot,
        },
        inputs: [
            `{projectRoot}/**/*.(${options.extensions.join('|')})`,
            ...inputConfigFiles,
            { externalDependencies: ['stylelint'] },
        ],
        metadata: {
            technologies: ['stylelint'],
            description: 'Runs Stylelint on project',
            help: {
                command: `${pmc.exec} stylelint --help`,
                example: {},
            },
        },
    };
}
function normalizeOptions(options) {
    let extensions;
    if (Array.isArray(options?.extensions)) {
        extensions = [...new Set(options.extensions.map((f) => f.replace(/^\.+/, '')))];
    }
    return {
        targetName: options?.targetName ?? 'stylelint',
        extensions: extensions ?? ['css'],
    };
}
